### PR TITLE
Get accurate model output max

### PR DESCRIPTION
**Pull request by pablodanswer**
## Description
Adjusts maximum input size based on the actual context window

## How Has This Been Tested?
- [x] Various model providers, context lengths, documents in input window

### CODE REVIEW

**Comment by pablodanswer** [comment]
I can think of a few options here (where `max_context` is total max context, `input_context` is the maximum input we set aside and `max_output` is the actual maximum output for a model)
1. `input_context` = `max_context` - `max_output` (not feasible)
2. `input_context` = max(`max_context` - `max_output`, `minimum variable set by an environment variable`) - leaving aside at least some space for the output (could also be an environment variable)
3. `input_context` =`max_context` - `max_output` * `multiplier < 1 set by an environment variable`

File: backend/danswer/llm/chat_llm.py [hunk 1]
```diff
@@ -228,6 +238,30 @@ def __init__(
     def log_model_configs(self) -> None:
         logger.debug(f"Config: {self.config}")
 
+    def _calculate_max_tokens(self, prompt: LanguageModelInput) -> int:
```
Comment (Weves) [inline backend/danswer/llm/chat_llm.py]: nit: renaming to `_calculate_max_output_tokens`. `max_tokens` is a bit ambiguous

File: backend/danswer/llm/chat_llm.py [hunk 2]
```diff
@@ -209,7 +211,15 @@ def __init__(
         self._api_base = api_base
         self._api_version = api_version
         self._custom_llm_provider = custom_llm_provider
-        self._max_output_tokens = max_output_tokens
+        self._max_output_tokens = (
```
Comment (Weves) [inline backend/danswer/llm/chat_llm.py]: this is not used either, right? If so, can we comment it + add brief comment as to why

File: backend/danswer/configs/model_configs.py [hunk 3]
```diff
@@ -80,11 +80,14 @@
 GEN_AI_LLM_PROVIDER_TYPE = os.environ.get("GEN_AI_LLM_PROVIDER_TYPE") or None
 # Override the auto-detection of LLM max context length
 GEN_AI_MAX_TOKENS = int(os.environ.get("GEN_AI_MAX_TOKENS") or 0) or None
+
 # Set this to be enough for an answer + quotes. Also used for Chat
+# This is the minimum token context we will leave for the LLM to generate an answer
 GEN_AI_MAX_OUTPUT_TOKENS = int(os.environ.get("GEN_AI_MAX_OUTPUT_TOKENS") or 1024)
```
Comment (Weves) [inline backend/danswer/configs/model_configs.py]: This should be renamed to `GEN_AI_NUM_RESERVED_OUTPUT_TOKENS` or something similar?

**Review by Weves — APPROVED** [review]

**Pull request merged**